test: make K8s DNS checks more robust

ianvernon opened the pull request:
This PR addresses flakiness in the CI that results from CoreDNS caching hostname entries for a long time. It addresses the following problem:

We have multiple tests which use the same YAML (`test/k8sT/manifests/demo_ds.yaml`) for creating resources in the CI:

* `test/k8sT/Tunnels.go`
* `test/k8sT/Chaos.go`
* `test/k8sT/Services.go`

These create services, which are referenced via hostnames in the CI. When CoreDNS is used in the CI (in Kubernetes >= 1.11 versions), hostname --> IP mappings are cached for some time, even after the service with the same name has been created in one test, and then re-created in another test shortly thereafter. This means that if any of the above tests run right after the other, there is a possibility that the IP returned by looking up a service name which was also used in the prior test which added a service with the same name would return the old IP for the service, which is not plumbed into the Cilium datapath, while the new service IP is. As a result, this means that tests which try to access services via hostname in the CI would fail because the DNS lookup would return the old IP for the service. This is odd behavior, considering that the `ConfigMap` for CoreDNS in the CI has the setting for cache (see https://coredns.io/plugins/cache/ ) to have a TTL of 3 seconds:

https://github.com/cilium/cilium/blob/1f7a845c49df4717bb1c62904abbaf5525a98cfb/test/provision/manifest/coredns_deployment.yaml#L88-L103

To alleviate the flakiness that results in the CI, this is addressed by the addition of functionality to wait for DNS entries to be ready by not only checking that `dig` returns an IP for the given service name, but that the IP that is returned by `dig` matches the `ClusterIP` for the service returned by Kubernetes. 

As part of debugging this, I also added some additional output / commands to make debugging such issues easier if they should arise in the future. The `WaitForKubeDNSEntry` helper function also runs `dig` for the given service name to provide additional information for debugging in the logs, and additional narration of the test in `test/k8sT/Chaos.go` to make it easier to determine how far the test has progressed while it runs. Finally, it also preforms a datapath check to not only the service name, but also the ClusterIP for the given service, to make it easier to see whether there is an issue with DNS resolution, or with the Cilium datapath itself in the test `test/k8sT/Chaos.go`.

Ideally, we should not see errors like the following in the CI any more after this is merged:

```
Cannot curl from "testclient-24crq" to testds-service
Expected command: kubectl exec -n default testclient-24crq -- curl -s -D /dev/stderr --fail --connect-timeout 3 --max-time 3 http://testds-service.default.svc.cluster.local:80/ 
To succeed, but it failed:
Exitcode: 28 
Stdout:
 	 
Stderr:
 	 command terminated with exit code 28
```

Signed-off by: Ian Vernon <ian@cilium.io>

Fixes: #5122 

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5133)
<!-- Reviewable:end -->

Commit:
test/k8sT: add more logs to narrate progress of ChaosTest

Also query the ClusterIP as well as the service name to make more clear whether
connectivity issues are purely due to DNS, or due to bugs in the Cilium
datapath.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
test/helpers: change parameters for WaitForKubeDNSEntry

Change function to accept service name and service namespace. Propagate this
change across the test suites.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
test: add additional WaitForKubeDNSEntry checks

If a service with the same name is installed, deleted, and installed again, it
is possible for DNS to cache the old service IP due to long TTL. As a result,
the lookup for an IP for DNS would succeed in `WaitForKubeDNSEntry`, because
the old entry was cached. This would result in the function returning and
indicating success to the caller, but tests using the service name would fail
because the IP that was returned was not the actual service IP. Add logic to
detect if the IP returned by looking up service names via DNS is the same as the
IP stored for the given service in Kubernetes, and use this as a condition for
blocking the return of `WaitForKubeDNSEntry`.

If the service is headless, then do not check for the service IP stored in
Kubernetes to match the IP returned for the service from kube-dns. This is
because the IP for headless services in Kubernetes is 'None', which will never
match the IP returned by kube-dns.

Signed-off by: Ian Vernon <ian@cilium.io>

eloycoto commented on test/helpers/kubectl.go:
```diff
@@ -626,41 +626,77 @@ func (kub *Kubectl) WaitKubeDNS() error {
+			res := kub.Exec(fmt.Sprintf(digCMD, serviceNameWithNamespace, dnsClusterIP))
+			serviceIPFromDNS := res.SingleOut()
+			isIP := govalidator.IsIP(serviceIPFromDNS)
+			if isIP {
```
I would change this if to something like 
```
if !isIP{
  return false
}
```

Is clear, and you avoid another tab for the following code. 
`

ianvernon replied:
Fixed in next upload.

eloycoto commented on test/helpers/kubectl.go:
```diff
@@ -626,41 +626,77 @@ func (kub *Kubectl) WaitKubeDNS() error {
+				if !strings.Contains(serviceIPFromDNS, serviceIP) {
+					logger.Debugf("service IP from DNS (%s) is not the IP for the service in Kubernetes (%s)\n", serviceIPFromDNS, serviceIP)
+					return false
+				} else {
```
If you do a return false before, you don't need to have this else.

ianvernon replied:
Refactored according to your comments elsewhere. Fixed in next upload.

eloycoto commented on test/helpers/kubectl.go:
```diff
@@ -626,41 +626,77 @@ func (kub *Kubectl) WaitKubeDNS() error {
+			isIP := govalidator.IsIP(serviceIPFromDNS)
+			if isIP {
+				if !strings.Contains(serviceIPFromDNS, serviceIP) {
+					logger.Debugf("service IP from DNS (%s) is not the IP for the service in Kubernetes (%s)\n", serviceIPFromDNS, serviceIP)
```
Service IP mismatch DNS returns '%s' and kubernetes has '%s'

ianvernon replied:
Edited the log message locally to denote that the IPs do not match, same for the case where they do match. Fixed in next upload.

eloycoto commented on test/helpers/kubectl.go:
```diff
@@ -626,41 +626,77 @@ func (kub *Kubectl) WaitKubeDNS() error {
+			isIP := govalidator.IsIP(serviceIPFromDNS)
+			if isIP {
+				if !strings.Contains(serviceIPFromDNS, serviceIP) {
+					logger.Debugf("service IP from DNS (%s) is not the IP for the service in Kubernetes (%s)\n", serviceIPFromDNS, serviceIP)
```
You don't need the `\n` at the end.

ianvernon replied:
Fixed in next upload.

eloycoto commented on test/helpers/kubectl.go:
```diff
@@ -626,41 +626,77 @@ func (kub *Kubectl) WaitKubeDNS() error {
+					return false
+				} else {
+					logger.Debugf("service IP from DNS (%s)  is the IP for the service in Kubernetes (%s)\n", serviceIPFromDNS, serviceIP)
+					doServicesMatch = true
```
If you return true, the function will finish and it'll not work..

ianvernon replied:
Good catch. I refactored the function in line with what you mentioned elsewhere on this PR.

eloycoto reviewed (dismissed):
TBH, I think that you made a over engineering in the body function for the DNS, I think that the logic is much easier if you do like this: 
```golang
body := func() bool {
	serviceIP, _, err := kub.GetServiceHostPort(serviceNamespace, serviceName)
	if err != nil {
		log.WithError(err).Errorf("cannot get service IP for service %s", serviceNameWithNamespace)
		return false
	}
	serviceIPFromDNS := kub.Exec(fmt.Sprintf(digCMD, serviceNameWithNamespace, dnsClusterIP)).SingleOut()
	if !govalidator.IsIP(serviceIPFromDNS){
			// Is not a valid IP, run this commands to log the information
			_ = kub.Exec(fmt.Sprintf(digCmdLong, serviceNameWithNamespace, dnsClusterIP))
			// TCP connection to know if is a connectivity issue.
			_ = kub.Exec(fmt.Sprintf(digCMDFallback, serviceNameWithNamespace, dnsClusterIP))
	}

	if !strings.Contains(serviceIPFromDNS, serviceIP) {
		logger.Debugf("service IP from DNS (%s) is not the IP for the service in Kubernetes (%s)\n", serviceIPFromDNS, serviceIP)
		return false
	}
}
```

eloycoto commented on test/k8sT/Chaos.go:
```diff
@@ -77,15 +77,26 @@ var _ = Describe("K8sChaosTest", func() {
 
+				By("\tGetting ClusterIP For testds-service")
+				host, _, err := kubectl.GetServiceHostPort(helpers.DefaultNamespace, "testds-service")
+				ExpectWithOffset(1, err).To(BeNil(), "unable to get ClusterIP and port for service testds-service")
```
TBH, I would delete all ExpectWithOffset in this function. When multiple Asserts on a function it a bit mess to know where fails.

ianvernon replied:
I disagree here, per our discussion offline.

eloycoto commented on test/k8sT/Chaos.go:
```diff
@@ -77,15 +77,26 @@ var _ = Describe("K8sChaosTest", func() {
 
-				err = kubectl.WaitForKubeDNSEntry(testDSService)
+				By("\tWaiting for kube-dns entry for service testds-service")
+				err = kubectl.WaitForKubeDNSEntry(testDSService, helpers.DefaultNamespace)
```
Can this be out of the loop? We tested for 6 different pods, and per 2. That can be out of the loop no?

(Worried about the time that this takes)

ianvernon replied:
Good idea.

ianvernon replied:
Will be fixed in next upload.

eloycoto commented on test/k8sT/Chaos.go:
```diff
@@ -77,15 +77,26 @@ var _ = Describe("K8sChaosTest", func() {
 				ExpectWithOffset(1, err).To(BeNil(), "DNS entry is not ready after timeout")
 
+				By("\tGetting ClusterIP For testds-service")
+				host, _, err := kubectl.GetServiceHostPort(helpers.DefaultNamespace, "testds-service")
```
That can be out of the loop too, is a global thing for all pods.

ianvernon replied:
Will be fixed in next upload.

Commit:
test/k8sT: move DNS checks outside of loop

Waiting for DNS entries to exist in kube-dns only needs to be done once per
invocation of `PingService` in this test, because once the entry exists, it will
persist; there is no need to check it within the loop for checking connectivity,
which runs multiple times.

Signed-off by: Ian Vernon <ian@cilium.io>

eloycoto commented on test/helpers/kubectl.go:
```diff
@@ -626,41 +626,79 @@ func (kub *Kubectl) WaitKubeDNS() error {
+		// ClusterIPNone denotes that this service is headless; there is no
+		// service IP for this service, and thus the IP returned by `dig` is
+		// an IP of the pod itself, not ClusterIPNone.
+		if serviceIP != v1.ClusterIPNone {
```
please, follow always the same format: 
```golang
if serviceIP == v1.ClusterIPNone {
	res := kub.Exec(fmt.Sprintf(digCMD, serviceNameWithNamespace, dnsClusterIP))
	if !res.WasSuccessful() {
		_ = kub.Exec(fmt.Sprintf(digCMDFallback, serviceNameWithNamespace, dnsClusterIP))
	}
	res.WasSuccessful() 
}
```

ianvernon replied:
I unconditionally want to run 

```
res := kub.Exec(fmt.Sprintf(digCMD, serviceNameWithNamespace, dnsClusterIP))
if !res.WasSuccessful() {
    _ = kub.Exec(fmt.Sprintf(digCMDFallback, serviceNameWithNamespace, dnsClusterIP))
}
res.WasSuccessful() 
```

regardless of it's a `ClusterIP` or not so I don't understand the rationale behind what you are saying here.

ianvernon replied:
Never mind, I already execute `digCmd`, you are right.

eloycoto commented on test/helpers/kubectl.go:
```diff
@@ -626,41 +626,79 @@ func (kub *Kubectl) WaitKubeDNS() error {
+				_ = kub.Exec(fmt.Sprintf(digCMDFallback, serviceNameWithNamespace, dnsClusterIP))
+				return false
+			}
+			logger.Debugf("service IP retrieved from DNS (%s) matches the IP for the service stored in Kubernetes (%s)", serviceIPFromDNS, serviceIP)
```
here you need a `return true`

ianvernon replied:
Fixed in next upload.

eloycoto commented on test/k8sT/Chaos.go:
```diff
@@ -75,17 +75,28 @@ var _ = Describe("K8sChaosTest", func() {
+
 		for _, pod := range pods {
 			for _, ip := range dsPods {
+				By("\tPinging test-ds service pod with IP %s from client pod %s", ip, pod)
```
please use %q to know that it's a variable when is print.

ianvernon replied:
Fixed in latest upload.

eloycoto commented on test/k8sT/Chaos.go:
```diff
@@ -75,17 +75,28 @@ var _ = Describe("K8sChaosTest", func() {
 
-				err = kubectl.WaitForKubeDNSEntry(testDSService)
-				ExpectWithOffset(1, err).To(BeNil(), "DNS entry is not ready after timeout")
+				By("\tCurling testds-service via ClusterIP %s", host)
```
Please use %q to know that it's a variable when is print

ianvernon replied:
Fixed in latest upload.

eloycoto commented on test/k8sT/Chaos.go:
```diff
@@ -75,17 +75,28 @@ var _ = Describe("K8sChaosTest", func() {
 		Expect(err).To(BeNil(), "Cannot get daemonset pods IPS")
 		Expect(len(pods)).To(BeNumerically(">", 0), "No pods available to test connectivity")
 
+		By("\tWaiting for kube-dns entry for service testds-service")
```
Due the place of this By, I don't think that the `\t` make sense here.

ianvernon replied:
I'll just remove the use of `\t` from `PingService`.

ianvernon replied:
Removed use of tabs.

eloycoto commented on test/k8sT/Chaos.go:
```diff
@@ -75,17 +75,28 @@ var _ = Describe("K8sChaosTest", func() {
+		err = kubectl.WaitForKubeDNSEntry(testDSService, helpers.DefaultNamespace)
+		ExpectWithOffset(1, err).To(BeNil(), "DNS entry is not ready after timeout")
+
+		By("\tGetting ClusterIP For testds-service")
```
The same here, does not make sense to me to have the \t here.

ianvernon replied:
Removed use of tabs.

eloycoto commented on test/helpers/kubectl.go:
```diff
@@ -626,41 +626,79 @@ func (kub *Kubectl) WaitKubeDNS() error {
+		}
+
+		res := kub.Exec(fmt.Sprintf(digCMD, serviceNameWithNamespace, dnsClusterIP))
+		_ = kub.Exec(fmt.Sprintf(digCmdLong, serviceNameWithNamespace, dnsClusterIP))
```
I don't think that the `digCMDLog` is needed at all. The full output is with the digCMDFallback, if you use here again there is no value added at all. I would delete it.

ianvernon replied:
OK; the reason I added `digCmdLog` was for when `dig` succeeded (an IP was found), but it wasn't the right IP for the service. We can just unconditionally run `digCmdFallback` in that case.

aanm reviewed (commented):
Nothing major on my side, I thought this PR touch agent code but it looks like it is only `test/`

eloycoto reviewed (approved):
now is super robust! :+1: Regards
